patroni/ha.py

```diff
@@ -447,27 +447,34 @@ def process_sync_replication(self):
         promoting standbys that were guaranteed to be replicating synchronously.
         """
         if self.is_synchronous_mode():
-            current = self.cluster.sync.leader and self.cluster.sync.sync_standby
-            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster)
-            if picked != current:
-                # We need to revoke privilege from current before replacing it in the config
-                if current:
-                    logger.info("Removing synchronous privilege from %s", current)
-                    if not self.dcs.write_sync_state(self.state_handler.name, None, index=self.cluster.sync.index):
+            sync_node_count = self.patroni.config['synchronous_node_count']
+            current = self.cluster.sync.leader and self.cluster.sync.members or []
+            picked, allow_promote = self.state_handler.pick_synchronous_standby(self.cluster, sync_node_count)
+            if set(picked) != set(current):
+                # update synchronous standby list in dcs temporarily to point to common nodes in current and picked
+                sync_common = list(set(current).intersection(set(allow_promote)))
+                if set(sync_common) != set(current):
+                    logger.info("Updating synchronous privilege temporarily from %s to %s", current, sync_common)
+                    if not self.dcs.write_sync_state(self.state_handler.name,
+                                                     sync_common or None,
+                                                     index=self.cluster.sync.index):
                         logger.info('Synchronous replication key updated by someone else.')
                         return
 
-                if self.is_synchronous_mode_strict() and picked is None:
-                    picked = '*'
+                # Update  db param and wait for x secs
+                if self.is_synchronous_mode_strict() and not picked:
+                    picked = ['*']
                     logger.warning("No standbys available!")
 
                 logger.info("Assigning synchronous standby status to %s", picked)
                 self.state_handler.config.set_synchronous_standby(picked)
 
-                if picked and picked != '*' and not allow_promote:
-                    # Wait for PostgreSQL to enable synchronous mode and see if we can immediately set sync_standby
+                if picked and not allow_promote and picked[0] != '*' and set(allow_promote) != set(picked):
```

Comment:
I mean if the `allow_promote` or `sync_common` (which effectively should be the same) are not empty, we can just return from this method.
Anything else should stay the same.

Reply:
got it. will fix it.

Reply:
I am thinking more of comparing `allow_promote` with `picked` and return if they are not same. This should optimize the call and let Patroni handle list in DCS during next iteration? Please let me know your thoughts.

```
                if (picked and picked[0] != '*' and set(allow_promote) != set(picked)):
                    logger.info("Skipping DCS Update for Patroni to take care in next iteration")
                    return
```

Reply:
We don't want to return if `allow_promote` is not empty and `picked` is empty, because it will leave the system in the state where only the former primary would be allowed to promote.

Reply:
I think this situation would never arise as the query in `[pick_synchronous_standby](https://github.com/ksarabu1/patroni/blob/feature/multisync/patroni/postgresql/__init__.py#L925-L927)` always return empty `allow_promote` if `picked` is empty.

Reply:
Sorry, I meant the opposite.
We don't want to return if `picked` is not empty and `allow_promote` is empty.

Reply:
Thanks. The following should mitigate that condition as well. Please let me know.

```
                if (picked and allow_promote and picked[0] != '*' and set(allow_promote) != set(picked)):
                    logger.info("Skipping DCS Update for Patroni to take care in next iteration")
                    return
```

Reply:
Reverting the logic back to previous [commit](https://github.com/zalando/patroni/commit/6a1523278e00597842aa804f2cdd18a130c54f13#diff-57c53e988b775ee2902488fb33c1f433L471) for now as it was tested and worked in all test cases.